Configure serializers/derserializers to be able to initialize Schema Registry

This is a follow up of Discord's short discussion with @iravid.

Configure serializers/deserializers to be able to initialize Schema Registry when using KafkaAvroSerilizer/KafkaAcroDeserializer.


While testing my producer against Kafka with Schema Registry it breaks down with following error:
```scala
[info]     org.apache.kafka.common.errors.SerializationException: Error serializing Avro message
[info]     Caused by: java.lang.NullPointerException
[info]          at io.confluent.kafka.serializers.AbstractKafkaAvroSerializer.serializeImpl(AbstractKafkaAvroSerializer.java:77)
[info]          at io.confluent.kafka.serializers.KafkaAvroSerializer.serialize(KafkaAvroSerializer.java:58)
[info]          at com.valamis.avro.serialization.AvroSerializerWithRegistry.$anonfun$serialize$1(AvroSerializerWithRegistry.scala:33)
[info]          at scala.Option.map(Option.scala:230)
[info]          at com.valamis.avro.serialization.AvroSerializerWithRegistry.serialize(AvroSerializerWithRegistry.scala:32)
[info]          at org.apache.kafka.common.serialization.Serializer.serialize(Serializer.java:62)
[info]          at zio.kafka.serde.Serializer$$anon$2.$anonfun$serialize$1(Serializer.scala:45)
[info]          at zio.internal.FiberContext.evaluateNow(FiberContext.scala:345)
[info]          at zio.internal.FiberContext.$anonfun$evaluateLater$1(FiberContext.scala:687)
[info]          at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[info]          at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[info]          at java.lang.Thread.run(Thread.java:748)
```

Basically,  Schema Registry hasn't gotten a chance to be initialized. In fact, it does that via method call 'configure' on Kafka Avro's serializers/deserializers. Have a look [here](https://github.com/confluentinc/schema-registry/blob/de6bfe255d92f22a67c72432338883b0da0ee82d/avro-serializer/src/main/java/io/confluent/kafka/serializers/AbstractKafkaAvroSerializer.java#L47) for further details. 

However, if you look at sources of `zio-kafka`'s Producer and Consumer there you will see that none of them call configure on underlining Kafka's Serializer/Deserializer. Hence, schema registry doesn't have a chance to be initialized while using Kavka Avro serdes.

If you  go through Scala Kafka client from Cake Solutions  and have a look at, say, Consumer then you will [see](https://github.com/cakesolutions/scala-kafka-client/blob/5f22aee5f48e6d1abe640d9316302733af56e68a/client/src/main/scala/cakesolutions/kafka/KafkaProducer.scala#L220) they call configure before do call serialize.

--- a/src/main/scala/zio/kafka/consumer/package.scala
+++ b/src/main/scala/zio/kafka/consumer/package.scala
@@ -174,18 +174,27 @@ package object consumer {
         Clock with Blocking,
         Throwable,
         (TopicPartition, ZStream[R, Throwable, CommittableRecord[K, V]])
-      ] =
-        ZStream
-          .fromQueue(runloop.partitions)
-          .collectWhileSuccess
-          .map {
-            case (tp, partition) =>
-              val partitionStream =
-                if (settings.perPartitionChunkPrefetch <= 0) partition
-                else partition.buffer(settings.perPartitionChunkPrefetch)
+      ] = {
 
-              tp -> partitionStream.mapM(_.deserializeWith(keyDeserializer, valueDeserializer))
+        val configureDeserializers =
+          ZStream.fromEffect {
+            keyDeserializer.configure(settings.driverSettings, isKey = true) *>
+              valueDeserializer.configure(settings.driverSettings, isKey = false)
           }
+
+        configureDeserializers *>
+          ZStream
+            .fromQueue(runloop.partitions)
+            .collectWhileSuccess
+            .map {
+              case (tp, partition) =>
+                val partitionStream =
+                  if (settings.perPartitionChunkPrefetch <= 0) partition
+                  else partition.buffer(settings.perPartitionChunkPrefetch)
+
+                tp -> partitionStream.mapM(_.deserializeWith(keyDeserializer, valueDeserializer))
+            }
+      }
 
       override def partitionsFor(
         topic: String,

--- a/src/main/scala/zio/kafka/producer/package.scala
+++ b/src/main/scala/zio/kafka/producer/package.scala
@@ -172,15 +172,18 @@ package object producer {
       keySerializer: Serializer[R, K],
       valueSerializer: Serializer[R, V]
     ): ZManaged[Any, Throwable, Service[R, K, V]] =
-      ZIO.effect {
-        val props = settings.driverSettings.asJava
-        val rawProducer = new KafkaProducer[Array[Byte], Array[Byte]](
-          props,
-          new ByteArraySerializer(),
-          new ByteArraySerializer()
-        )
-        Live(rawProducer, settings, keySerializer, valueSerializer)
-      }.toManaged(_.close)
+      (for {
+        props <- ZIO.effect(settings.driverSettings)
+        _     <- keySerializer.configure(props, isKey = true)
+        _     <- valueSerializer.configure(props, isKey = false)
+        rawProducer <- ZIO.effect(
+                        new KafkaProducer[Array[Byte], Array[Byte]](
+                          props.asJava,
+                          new ByteArraySerializer(),
+                          new ByteArraySerializer()
+                        )
+                      )
+      } yield Live(rawProducer, settings, keySerializer, valueSerializer)).toManaged(_.close)
 
     def withProducerService[R: Tag, K: Tag, V: Tag, A](
       r: Producer.Service[R, K, V] => RIO[R with Blocking, A]

--- a/src/main/scala/zio/kafka/serde/Deserializer.scala
+++ b/src/main/scala/zio/kafka/serde/Deserializer.scala
@@ -6,6 +6,7 @@ import zio.{ RIO, Task, ZIO }
 
 import scala.annotation.nowarn
 import scala.util.{ Failure, Success, Try }
+import scala.jdk.CollectionConverters._
 
 /**
  * Deserializer from byte array to a value of some type T
@@ -15,6 +16,7 @@ import scala.util.{ Failure, Success, Try }
  */
 trait Deserializer[-R, +T] {
   def deserialize(topic: String, headers: Headers, data: Array[Byte]): RIO[R, T]
+  def configure(props: Map[String, AnyRef], isKey: Boolean): Task[Unit]
 
   /**
    * Create a deserializer for a type U based on the deserializer for type T and a mapping function
@@ -56,6 +58,8 @@ object Deserializer extends Serdes {
   def apply[R, T](deser: (String, Headers, Array[Byte]) => RIO[R, T]): Deserializer[R, T] = new Deserializer[R, T] {
     override def deserialize(topic: String, headers: Headers, data: Array[Byte]): RIO[R, T] =
       deser(topic, headers, data)
+
+    override def configure(props: Map[String, AnyRef], isKey: Boolean): Task[Unit] = Task.unit
   }
 
   /**
@@ -64,5 +68,8 @@ object Deserializer extends Serdes {
   def apply[T](deserializer: KafkaDeserializer[T]): Deserializer[Any, T] = new Deserializer[Any, T] {
     override def deserialize(topic: String, headers: Headers, data: Array[Byte]): Task[T] =
       Task(deserializer.deserialize(topic, headers, data))
+
+    override def configure(props: Map[String, AnyRef], isKey: Boolean): zio.Task[Unit] =
+      Task(deserializer.configure(props.asJava, isKey))
   }
 }

--- a/src/main/scala/zio/kafka/serde/Serde.scala
+++ b/src/main/scala/zio/kafka/serde/Serde.scala
@@ -1,10 +1,12 @@
 package zio.kafka.serde
 
 import org.apache.kafka.common.serialization.{ Serde => KafkaSerde }
+import org.apache.kafka.common.header.Headers
+
 import zio.{ RIO, Task }
 
 import scala.util.Try
-import org.apache.kafka.common.header.Headers
+import scala.jdk.CollectionConverters._
 
 /**
  * A serializer and deserializer for values of type T
@@ -45,6 +47,7 @@ object Serde extends Serdes {
         ser(topic, headers, value)
       override def deserialize(topic: String, headers: Headers, data: Array[Byte]): RIO[R, T] =
         deser(topic, headers, data)
+      override def configure(props: Map[String, AnyRef], isKey: Boolean): Task[Unit] = Task.unit
     }
 
   /**
@@ -55,6 +58,8 @@ object Serde extends Serdes {
       ser.serialize(topic, headers, value)
     override def deserialize(topic: String, headers: Headers, data: Array[Byte]): RIO[R, T] =
       deser.deserialize(topic, headers, data)
+    override def configure(props: Map[String, AnyRef], isKey: Boolean): Task[Unit] =
+      deser.configure(props, isKey) *> ser.configure(props, isKey)
   }
 
   /**
@@ -65,6 +70,8 @@ object Serde extends Serdes {
       Task(serde.serializer().serialize(topic, headers, value))
     override def deserialize(topic: String, headers: Headers, data: Array[Byte]): Task[T] =
       Task(serde.deserializer().deserialize(topic, headers, data))
+    override def configure(props: Map[String, AnyRef], isKey: Boolean): Task[Unit] =
+      Task(serde.configure(props.asJava, isKey))
   }
 
   implicit def deserializerWithError[R, T](implicit deser: Deserializer[R, T]): Deserializer[R, Try[T]] =

--- a/src/main/scala/zio/kafka/serde/Serializer.scala
+++ b/src/main/scala/zio/kafka/serde/Serializer.scala
@@ -4,6 +4,8 @@ import zio.{ RIO, Task }
 import org.apache.kafka.common.header.Headers
 import org.apache.kafka.common.serialization.{ Serializer => KafkaSerializer }
 
+import scala.jdk.CollectionConverters._
+
 /**
  * Serializer from values of some type T to a byte array
  *
@@ -12,6 +14,7 @@ import org.apache.kafka.common.serialization.{ Serializer => KafkaSerializer }
  */
 trait Serializer[-R, -T] {
   def serialize(topic: String, headers: Headers, value: T): RIO[R, Array[Byte]]
+  def configure(props: Map[String, AnyRef], isKey: Boolean): Task[Unit]
 
   /**
    * Create a serializer for a type U based on the serializer for type T and a mapping function
@@ -38,6 +41,8 @@ object Serializer extends Serdes {
     new Serializer[R, T] {
       override def serialize(topic: String, headers: Headers, value: T): RIO[R, Array[Byte]] =
         ser(topic, headers, value)
+
+      override def configure(props: Map[String, AnyRef], isKey: Boolean): Task[Unit] = Task.unit
     }
 
   /**
@@ -46,5 +51,8 @@ object Serializer extends Serdes {
   def apply[T](serializer: KafkaSerializer[T]): Serializer[Any, T] = new Serializer[Any, T] {
     override def serialize(topic: String, headers: Headers, value: T): Task[Array[Byte]] =
       Task(serializer.serialize(topic, headers, value))
+
+    override def configure(props: Map[String, AnyRef], isKey: Boolean): Task[Unit] =
+      Task(serializer.configure(props.asJava, isKey))
   }
 }

